sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -1408,11 +1409,37 @@ case class ValidateExternalType(child: Expression, expected: DataType)
 
   override def dataType: DataType = RowEncoder.externalDataTypeForInput(expected)
 
-  override def eval(input: InternalRow): Any =
-    throw new UnsupportedOperationException("Only code-generated evaluation is supported")
-
   private val errMsg = s" is not a valid external type for schema of ${expected.simpleString}"
 
+  private lazy val checkType = expected match {
+    case _: DecimalType =>
+      (value: Any) => {
+        Seq(classOf[java.math.BigDecimal], classOf[scala.math.BigDecimal], classOf[Decimal])
+          .exists { x => value.getClass.isAssignableFrom(x) }
+      }
+    case _: ArrayType =>
+      (value: Any) => {
+        value.getClass.isAssignableFrom(classOf[Seq[_]]) || value.getClass.isArray
+      }
+    case _ if ScalaReflection.isNativeType(expected) =>
+      (value: Any) => {
+        value.getClass.isAssignableFrom(ScalaReflection.classForNativeTypeOf(expected))
```

Comment:
Let's move the `ScalaReflection.classForNativeTypeOf(expected)` out of the function body and into a `val`.

Reply:
It might also be a bit faster if we use `isInstance` here. All of these classes do not have subclasses.

Reply:
ok